File: src/main/java/com/mudhut/nudge/users/services/VerificationService.java
package com.mudhut.nudge.users.services;

import java.time.LocalDateTime;
import java.util.Random;
import java.util.UUID;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.mudhut.nudge.email.JavaEmailService;
import com.mudhut.nudge.users.entities.PhoneVerificationToken;
import com.mudhut.nudge.users.entities.User;
import com.mudhut.nudge.users.entities.VerificationToken;
import com.mudhut.nudge.users.repositories.PhoneVerificationTokenRepository;
import com.mudhut.nudge.users.repositories.UserRepository;
import com.mudhut.nudge.users.repositories.VerificationTokenRepository;
import com.mudhut.nudge.utils.UrlService;

@Service
public class VerificationService {

    @Autowired
    VerificationTokenRepository verificationTokenRepository;

    @Autowired
    PhoneVerificationTokenRepository phoneVerificationTokenRepository;

    @Autowired
    UserRepository userRepository;

    @Autowired
    private JavaEmailService emailService;

    @Autowired
    private UrlService urlService;

    private String generateRandomToken() {
        return UUID.randomUUID().toString();
    }

    // Helper method to generate and save verification token
    public String createVerificationToken(User user) {
        String token = generateRandomToken(); // Implement this method to generate a random token
        VerificationToken verificationToken = new VerificationToken(token, user);
        verificationTokenRepository.save(verificationToken);
        return token;
    }

    private String generateVerificationCode() {
        // Generate a 6-digit code
        return String.format("%06d", new Random().nextInt(1000000));
    }

    // Helper method to generate and save phone verification code
    public String createPhoneVerificationCode(User user) {
        String code = generateVerificationCode();
        PhoneVerificationToken verificationToken = new PhoneVerificationToken(code, user);
        phoneVerificationTokenRepository.save(verificationToken);
        return code;
    }

    public void sendVerificationEmail(User user, String token) {
        // Create verification URL
        String verificationUrl = urlService.buildUrlWithParam("/verify-email", "token", token);

        // Email content
        String subject = "Please verify your email address";

        StringBuilder contentBuilder = new StringBuilder();

        contentBuilder.append("Dear User,\n\n")
                .append("Please click the link below to verify your email address:\n")
                .append(verificationUrl).append("\n\n")
                .append("This link will expire in 24 hours.\n\n")
                .append("Thank you,\n")
                .append("The Nudge App Team");

        String content = contentBuilder.toString();

        // Send the email
        emailService.sendEmail(user.getEmail(), subject, content);
    }

    @Transactional
    public void verifyEmail(String token) {
        VerificationToken verificationToken = verificationTokenRepository.findByToken(token)
                .orElseThrow(() -> new IllegalArgumentException("Invalid verification token"));

        if (verificationToken.isUsed()) {
            throw new IllegalStateException("Token has already been used");
        }

        if (verificationToken.getExpiryDate().isBefore(LocalDateTime.now())) {
            throw new IllegalStateException("Token has expired");
        }

        User user = verificationToken.getUser();
        user.setEmailVerified(true);
        user.setActive(true); // Activate user account upon email verification

        verificationToken.setUsed(true);

        userRepository.save(user);
        verificationTokenRepository.save(verificationToken);
    }

    @Transactional
    public void verifyPhone(String code) {
        PhoneVerificationToken verificationToken = phoneVerificationTokenRepository.findByCode(code)
                .orElseThrow(() -> new IllegalArgumentException("Invalid verification code"));

        if (verificationToken.isUsed()) {
            throw new IllegalStateException("Code has already been used");
        }

        if (verificationToken.getExpiryDate().isBefore(LocalDateTime.now())) {
            throw new IllegalStateException("Code has expired");
        }

        User user = verificationToken.getUser();
        user.setPhoneVerified(true);

        // If email is already verified, activate the account
        if (user.isEmailVerified()) {
            user.setActive(true);
        }

        verificationToken.setUsed(true);

        userRepository.save(user);
        phoneVerificationTokenRepository.save(verificationToken);
    }
}

File: src/main/kotlin/com/mudhut/nudge/users/services/VerificationService.kt
package com.mudhut.nudge.users.services

import com.mudhut.nudge.email.IEmailService
import com.mudhut.nudge.users.entities.PhoneVerificationToken
import com.mudhut.nudge.users.entities.User
import com.mudhut.nudge.users.entities.VerificationToken
import com.mudhut.nudge.users.repositories.PhoneVerificationTokenRepository
import com.mudhut.nudge.users.repositories.UserRepository
import com.mudhut.nudge.users.repositories.VerificationTokenRepository
import com.mudhut.nudge.utils.UrlService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.thymeleaf.TemplateEngine
import org.thymeleaf.context.Context
import java.time.LocalDateTime
import java.util.*

private const val VERIFICATION_EMAIL_SUBJECT = "Confirm your email to get started"

@Service
class VerificationService(
    private val verificationTokenRepository: VerificationTokenRepository,
    private val phoneVerificationTokenRepository: PhoneVerificationTokenRepository,
    private val userRepository: UserRepository,
    private val emailService: IEmailService,
    private val urlService: UrlService,
    private val templateEngine: TemplateEngine,
) {

    private fun generateRandomToken(): String = UUID.randomUUID().toString()

    fun createVerificationToken(user: User): String {
        val token = generateRandomToken()
        val verificationToken = VerificationToken(token, user)
        verificationTokenRepository.save(verificationToken)
        return token
    }

    private fun generateVerificationCode(): String =
        String.format("%06d", Random().nextInt(1000000))

    fun createPhoneVerificationCode(user: User): String {
        val code = generateVerificationCode()
        val verificationToken = PhoneVerificationToken(code, user)
        phoneVerificationTokenRepository.save(verificationToken)
        return code
    }

    fun sendVerificationEmail(user: User, token: String) {
        val verificationUrl = urlService.buildUrlWithParam("/verify-email", "token", token)
        val displayName = user.username ?: "there"

        val context = Context().apply {
            setVariable("displayName", displayName)
            setVariable("verificationUrl", verificationUrl)
            setVariable("subject", VERIFICATION_EMAIL_SUBJECT)
        }
        val html = templateEngine.process("emails/verification", context)
        val text = plainTextVerificationEmail(displayName, verificationUrl)

        emailService.sendHtmlEmail(user.email!!, VERIFICATION_EMAIL_SUBJECT, html, text)
    }

    private fun plainTextVerificationEmail(displayName: String, verificationUrl: String): String =
        """
        Hi $displayName,

        Thanks for signing up for Nudge. Click the link below to verify your email address and activate your account:

        $verificationUrl

        This link expires in 24 hours.

        If you didn't sign up for Nudge, you can safely ignore this email.

        — The Nudge team
        """.trimIndent()

    @Transactional
    fun verifyEmail(token: String) {
        val verificationToken = verificationTokenRepository.findByToken(token)
            .orElseThrow { IllegalArgumentException("Invalid verification token") }

        if (verificationToken.used) {
            throw IllegalStateException("Token has already been used")
        }

        if (verificationToken.expiryDate!!.isBefore(LocalDateTime.now())) {
            throw IllegalStateException("Token has expired")
        }

        val user = verificationToken.user!!
        user.isEmailVerified = true
        user.isActive = true

        verificationToken.used = true

        userRepository.save(user)
        verificationTokenRepository.save(verificationToken)
    }

    @Transactional
    fun verifyPhone(code: String) {
        val verificationToken = phoneVerificationTokenRepository.findByCode(code)
            .orElseThrow { IllegalArgumentException("Invalid verification code") }

        if (verificationToken.used) {
            throw IllegalStateException("Code has already been used")
        }

        if (verificationToken.expiryDate!!.isBefore(LocalDateTime.now())) {
            throw IllegalStateException("Code has expired")
        }

        val user = verificationToken.user!!
        user.isPhoneVerified = true

        if (user.isEmailVerified) {
            user.isActive = true
        }

        verificationToken.used = true

        userRepository.save(user)
        phoneVerificationTokenRepository.save(verificationToken)
    }
}
